Source/WebDriver/WebSocketServer.h

```diff
@@ -0,0 +1,146 @@
+    RefPtr<JSON::Object> parameters;
+
+    static std::optional<Command> fromData(const char* data, size_t dataLength);
+};
+
+// https://w3c.github.io/webdriver-bidi/#websocket-listener
+// A WebSocket listener is a network endpoint that is able to accept incoming WebSocket connections.
+// A WebSocket listener has a host, a port, a secure flag, and a list of WebSocket resources.
+struct WebSocketListener : public RefCounted<WebSocketListener> {
+    String host;
+    unsigned port;
+    bool secure;
+    std::vector<String> resources;
+
+    static RefPtr<WebSocketListener> create(const String& host, unsigned port, bool secure, const std::vector<String>& resources = { })
+    {
+        return adoptRef(*new WebSocketListener(host, port, secure, resources));
+    }
+
+private:
+    // Constructor is private to ensure it can only be called from within the class, specifically from the create method.
+    WebSocketListener(const String& host, unsigned port, bool secure, const std::vector<String>& resources)
+        : host(host), port(port), secure(secure), resources(resources) { }
+};
+
+
+class WebSocketMessageHandler {
+public:
+
+#if USE(SOUP)
+    typedef GRefPtr<SoupWebsocketConnection> Connection;
+#endif
+
+    struct Message {
+        // Optional connection, as the message might be a reply to a command not yet associated to a connection
+        std::optional<Connection> connection;
+        const char* data { nullptr };
+        size_t dataLength { 0 };
+
+        static Message fail(CommandResult::ErrorCode, std::optional<Connection>, std::optional<String> errorMessage = std::nullopt, std::optional<String> commandId = std::nullopt);
+    };
+
+    virtual void handleHandshake(HTTPRequestHandler::Request&&, Function<void(std::optional<HTTPRequestHandler::Response>&& response)>&& replyHandler) = 0;
+    virtual void handleMessage(Message&&, Function<void(Message&&)>&& replyHandler) = 0;
+    virtual void clientDisconnected(Connection) = 0;
+private:
+};
+
+class WebSocketServer : public RefCounted<WebSocketServer>, public CanMakeWeakPtr<WebSocketServer> {
+public:
+    explicit WebSocketServer(WebSocketMessageHandler&, WebDriverService&);
+    virtual ~WebSocketServer() = default;
+
+    std::optional<String> listen(const std::optional<String>& host, unsigned port);
+    bool isListening();
+    void disconnect();
+
+    WebSocketMessageHandler& messageHandler() { return m_messageHandler; }
+
+    const RefPtr<WebSocketListener> listener() const { return m_listener; }
+    void addToConnectionsNotAssociatedToASession(WebSocketMessageHandler::Connection);
+    void addToConnectionsAssociatedToASession(std::pair<WebSocketMessageHandler::Connection, String>);
+    bool isConnectionNotAssociatedToSession(WebSocketMessageHandler::Connection&);
+    void removeConnectionNotAssociatedWithSession(WebSocketMessageHandler::Connection&);
+    std::optional<RefPtr<Session>> getSessionForConnection(WebSocketMessageHandler::Connection&);
```

Comment:
Naming: many of the method names are redundant with the argument names, so I think they could be shortened without losing any information.

Reply:
Renamed